### ISSUE TITLE
questiion about built-in model in gnina

### ISSUE THREAD

**Issue by croshong** [gnina/gnina#193]
it seems there are following built-in 3D cnn model which can be used when I
run gnina
crossdock_default2018
crossdock_default2018_1
crossdock_default2018_2
crossdock_default2018_3
crossdock_default2018_4
default2017
dense
dense_1
dense_2
dense_3
dense_4
general_default2018
general_default2018_1
general_default2018_2
general_default2018_3
general_default2018_4
redock_default2018
redock_default2018_1
redock_default2018_2
redock_default2018_3
redock_default2018_4

I have three question about using model
1) I understand that crossdock_default2018~~ means crossdock data trained with default2018 model
but what does *_1,*_2,*_3,*_4 means? Does it mean clustered cross validated model?
2) When I use each of the above model throu gnina binary with --cnn option, is the predicted
CNNscore, CNNaffinity from 3 fold model average ?
3) Does the general_default2018 means PDB_Bind General dataset trained with default2018 model?
the reason that I ask this question is in github,there are 2 kinds of general dataset present,
one is https://github.com/gnina/models/tree/master/data/general, which was published in 2017
the other is https://github.com/gnina/models/tree/master/data/PDBBind2016, which was published in 2020 with crossdock data

Thanks for your kind answer

**Comment by dkoes** [comment]
All the models are trained on the full dataset.  The different numbers are models trained with different random seeds.  I'll let @francoep answer the general set question.

**Comment by francoep** [comment]
The built-in models were trained with the  [ https://github.com/gnina/models/tree/master/data/PDBBind2016](https://github.com/gnina/models/tree/master/data/PDBBind2016) dataset, which were used for the data with the CrossDocked2020 paper

**Comment by croshong** [comment]
Thanks for your answer , I understand the all models are trained with full data set not the clustered cross validated set, To make it more clear, what you mean by full data is really whole data set?  not the 3 fold validated splited data ?

**Comment by francoep** [comment]
Correct. The full dataset means all of the data that is utilized the clustered crossvalidation. The types files for the complete set of data are provided. The numbers in the different files refer to the random seed that was used to train them.

**Comment by croshong** [comment]
Yes I understand 
What you mean is that the PDBbind based in built model in gnina( general_default2018, general_default2018_1, 
general_default2018_2, general_default2018_3,general_default2018_4) 
was trained with gen_uff_train0.types/gen_uff_train0.types in https://github.com/gnina/models/tree/master/data/PDBBind2016/General_types
and default2018.model in https://github.com/gnina/models/tree/master/crossdocked_paper using train.py in https://github.com/gnina/scripts?

If that is correct, I have question 

I made model with  from gen_uff_train0.types/gen_uff_train0.types with default2018.model using train.py use it to predict
CNNaffinity, CNNScore of py pdb complex and compare with the result in gnina in built model

1. my model built from gen_uff_train0.types/gen_uff_train0.types with default2018.model
command line 
gnina -r rec.pdb -l het.pdb --device 0 --cnn_model default2018.model --cnn_weights gen_uff_default2018.0_iter_80000.caffemodel --score_only
output
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.38091 
CNNaffinity: 6.98595
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000

2.gnina in built model general_default2018
command line
output
gnina -r rec.pdb -l het.pdb --device 0 --cnn general_default2018  --score_only
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.24911 
CNNaffinity: 7.36720
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000

3. gnina in built model general_default2018_1
command line 
gnina -r rec.pdb -l het.pdb --device 0 --cnn general_default2018_1  --score_only
output
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.55887 
CNNaffinity: 7.36869
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000

4. gnina in built model general_default2018_2
command line 
gnina -r rec.pdb -l het.pdb --device 0 --cnn general_default2018_2  --score_only
output
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.16036 
CNNaffinity: 6.84679
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000

5. gnina in built model general_default2018_3
command line 
gnina -r rec.pdb -l het.pdb --device 0 --cnn general_default2018_3  --score_only
output
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.33990 
CNNaffinity: 7.40911
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000


6.gnina in built model general_default2018_4
command line 
gnina -r rec.pdb -l het.pdb --device 0 --cnn general_default2018_4  --score_only
output
Affinity: -8.40892 (kcal/mol)
CNNscore: 0.07875 
CNNaffinity: 7.02902
Intramolecular energy: 7.11103
Term values, before weighting:
##  95.12461 1873.38940 1.92694 32.80760 0.43942 0.00000

I think that general_default, _1,_2,_3,_4 was trained with same data but with different random seed and may be my trained model 
should have different random seed,  In my opinion, the result can not be same but should be similar
but as you can see in above, I think the difference seems to be somewhat large.
Is there anything that I misunderstand or this difference is reasonable?
Thanks

**Comment by croshong** [comment]
one more simple question is that when I run gnina without any -cnn argument like

gnina -r rec.pdb -l het.pdb --device 0 --score_only

Is crossdock_default2018 the default in built model ?

**Comment by dkoes** [comment]
The default is to use an ensemble of 5 models (dense, general_default2018_3, dense_3, crossdock_Default2018, redock_default2018_2).  How these were chosen is described in the GNINA 1.0 paper.

**Comment by francoep** [comment]
The training of the general_default2018 models was done with the *_completeset_*.types file in that directory, which is not the same data that you trained your version of a model with. 

From training, we also observe a fair bit of variation between the different seeds of the model.

**Comment by croshong** [comment]
> The training of the general_default2018 models was done with the __completeset__.types file in that directory, which is not the same data that you trained your version of a model with.
> 
> From training, we also observe a fair bit of variation between the different seeds of the model.

What you mean by completeset.types is fixed_gen_uff_completeset_train0/test0.types?
what does it mean by 'fixed_' ?
and in the README.md file, it is written that
Train: Refined+Core Test:Refined+Core with Docked poses -- fixed_ref_uff_completeset_* ( maybe it should be fixed_gen_uff...)
that means for fixed_gen_uff_completeset, train set and test set is completely same.

As far as I know, train set and test set should be different/
Is there any other reason for setting train and test data same?

**Comment by croshong** [comment]
Also I understand there should be variation in model with different seeds, then when I predict the CNNaffinity, CNNscore with general_default2018, should I predict with each of 5 model(general_default2018, general_default2018_1, general_default2018_2
general_default2018_3,general_default2018_4) and take the mean value of 5 predicted value?

**Comment by francoep** [comment]
The completeset is every single data point. This the entire PDBBind general set (which is a super set containing all of the general set exclusive complexes, refined set complexes, and core set complexes).

The fixed prefix is because there were problems with some other versions of the dataset files, and in fixing them I used a different prefix so I could have the old files (that were wrong) without overwriting them.

The ref prefix is for the REFined set, which is a smaller more curated subset of the PDBBind. 

When you are training on all of the available data (which you do on every deployed model since it makes no sense to release a model for production purposes that is not trained on all your data points), there is no separate testing data. Hence they are the same (needed duplicate files for how some of our other scripts are set up).

Taking the output of 5 models trained with different seeds and taking their average is an ensemble model. We have shown in our work that this produces better results than taking the predictions of a single model. The downside is that this is more computationally demanding. But if you have the time/resources, I would recommend doing so.

**Comment by croshong** [comment]
What you mean is that 
fixed_ref_uff_completeset_* is the train/test datafile used in in built model of gnina 
and it is just entire PDBbind general set?

The ref prefix is for the REFined set, which is a smaller more curated subset of the PDBBind
-> Yes I know that but in the general set data explantation part of README.md of https://github.com/gnina/models/tree/master/data/PDBBind2016
     It is written that 
      General_types
      Clustered Cross Validated Splits                        -- ccv_gen_uff*
      Clustered Cross Validated Splits    Ligand Only         -- ccv_gen_norec_uff*
      Train: Refined-Core Test:Core  with Docked poses        -- gen_uff_*
      Train: Refined+Core Test:Refined+Core with Docked poses -- fixed_ref_uff_completeset_*
      I think Refined should be corrected to general 

When you are training on all of the available data 
(which you do on every deployed model since it makes no sense to release a model for production purposes that is not trained on all your data points), 
there is no separate testing data. Hence they are the same (needed duplicate files for how some of our other scripts are set up)
-> That means for general_default2018 model was trained with fixed_ref_uff_completeset_train0.types 
     but you made the duplicate file because some of your script ( such as train.py) are set up to work wth train.types, test.types
     which means that for PDBbind general default2018 model, because optimization for best result is finished, whole of data 
     was used for model generation for production purpose, but train.py( which is used in model generation) require train0.types, test0.types
      you made same file and use both of them in model generation script( train.py)

is there something wrong with my understanding?

Thanks for your patience for trying to understand my poor english..

**Comment by francoep** [comment]
Ah, right. Good catch on the mistake in the readme. Fixed that.  It was supposed to be fixed_gen_uff_completeset_train0.types.

The provided models in gnina are trained with different datasets. general_default_2018 is the pdbbind general set (fixed_gen_uff_completeset_train0.types), crossdock_* and dense_* is the complete CrossDocked2020 dataset, and redocked_* is the redocked subset of CrossDocked2020. I forget which set was used with default2017 (I think it is the pdbbind general set?) but that model is the worst of the provided ones so I wouldn't care too much.

So depending on what you want to compare to, you'll need to use a different weights file, and a different types file setup. For example, to compare to the deployed version of gnina's general_default2018 model, you would use said builtin model, and train your own of the fixed_gen_uff_completeset_* and evaluate.

If instead you wanted to compare to results in some of our papers (like earlier figures in the CrossDocked2020 publication), then you would need to use a def2018 model file with one of the other provided weights, along with a different types file setup for training.

**Comment by croshong** [comment]
The model training script train.py in https://github.com/gnina/scripts
require both of train.types and test.types 
to train general_default2018.model with train.py, should I use 
fixed_gen_uff_completeset_train0.types, fixed_gen_uff_completeset_test0.types
even two files are same because it is for production purpose?

**Comment by francoep** [comment]
I don't see the point of doing so, since that is just reproducing the built in model (assuming you used the same hyperparameter training settings that we used in our paper). But yes, that would be the approach.

If you are trying to compare between two different models, I would suggest utilizing a different training and testing approach and actually utilize a data split to be able to compare between them better.  But this entirely depends on what you are doing/experimenting with.
